File: Data Structures & Algorithms/string-encode-and-decode/submission-1.java
class Solution {
    int[] arr;

    public String encode(List<String> strs) {
        this.arr = new int[strs.size()];
        int i = 0;
        StringBuilder sb = new StringBuilder();
        for (String s : strs) {
            this.arr[i] = s.length();
            sb.append(s);
            i++;
        }
        return sb.toString();
    }

    public List<String> decode(String str) {
        List<String> res = new ArrayList<>();
        int s = 0;
        for (int n : this.arr) {
            if (n == 0) {
                res.add("");
            } else {
                int end = s + n;
                res.add(str.substring(s, end));
                s = end;
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/minimum-array-end/submission-1.py
class Solution:
    def minEnd(self, n: int, x: int) -> int:
        res = x
        bit_x = 1
        bit_n = 1

        while bit_n <= n - 1:
            if bit_x & x == 0:
                if bit_n & (n - 1):
                    res |= bit_x
                bit_n = bit_n << 1
            bit_x = bit_x << 1
        
        return res
